File: apps/windows/TeslaSync.App/feature-views/SignalCompareControls/SignalCompareControls.Model.cs
using System.Globalization;
using System.Text.RegularExpressions;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.FeatureViews.SignalDiff;

/// <summary>
/// One of the eight signal categories the compare bar exposes as a quick filter chip — the native port of an
/// entry in the web <c>CATEGORY_PREFIXES</c> array
/// (<c>web/src/features/telemetry/components/SignalCompareControls.tsx</c>). Carries the stable <see cref="Id"/>
/// (the wire value the page turns into a server-side filter prefix), the i18n <see cref="LabelKey"/> + English
/// <see cref="DefaultLabel"/>, and the case-insensitive name <see cref="Matches(string)"/> predicate the web
/// uses to decide whether a signal name belongs to the category. Pure (no WinUI types) so the matching rules
/// are asserted headlessly.
/// </summary>
public sealed class SignalCategory
{
    private readonly Regex _pattern;

    /// <summary>Creates a category over its id, i18n key, English fallback and the web name-match pattern.</summary>
    /// <param name="id">Stable wire id (web <c>id</c>), e.g. <c>battery</c>.</param>
    /// <param name="labelKey">Web i18n key (web <c>labelKey</c>), e.g. <c>signalDiff.cat.battery</c>.</param>
    /// <param name="defaultLabel">English fallback label (web <c>defaultLabel</c>).</param>
    /// <param name="pattern">The web regular-expression source the chip matches signal names against.</param>
    public SignalCategory(string id, string labelKey, string defaultLabel, string pattern)
    {
        ArgumentException.ThrowIfNullOrEmpty(id);
        ArgumentException.ThrowIfNullOrEmpty(labelKey);
        ArgumentException.ThrowIfNullOrEmpty(defaultLabel);
        ArgumentException.ThrowIfNullOrEmpty(pattern);

        Id = id;
        LabelKey = labelKey;
        DefaultLabel = defaultLabel;
        Pattern = pattern;
        _pattern = new Regex(pattern, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
    }

    /// <summary>Stable wire id (web <c>id</c>).</summary>
    public string Id { get; }

    /// <summary>Web i18n key (web <c>labelKey</c>).</summary>
    public string LabelKey { get; }

    /// <summary>English fallback label (web <c>defaultLabel</c>).</summary>
    public string DefaultLabel { get; }

    /// <summary>The web regular-expression source (kept for parity assertions).</summary>
    public string Pattern { get; }

    /// <summary>True when <paramref name="name"/> matches the category pattern (web <c>matches(name)</c>).</summary>
    public bool Matches(string? name) => !string.IsNullOrEmpty(name) && _pattern.IsMatch(name);
}

/// <summary>
/// The eight signal categories the compare bar offers, in display order — the native port of the web
/// <c>CATEGORY_PREFIXES</c> export. The pages reuse this list (and each entry's <see cref="SignalCategory.Matches"/>
/// predicate) to drive their server-side filter strings, exactly as the web does. Pure — unit-tested without a UI
/// host.
/// </summary>
public static class SignalCompareControlsCategories
{
    /// <summary>The eight categories in web declaration order.</summary>
    public static IReadOnlyList<SignalCategory> All { get; } = new[]
    {
        new SignalCategory("battery", "signalDiff.cat.battery", "Battery", "battery|charge|soc|range|kwh"),
        new SignalCategory("drive", "signalDiff.cat.drive", "Drive", "speed|odometer|gear|drive|brake|throttle|steering"),
        new SignalCategory("climate", "signalDiff.cat.climate", "Climate", "climate|hvac|cabin|seat|temp"),
        new SignalCategory("security", "signalDiff.cat.security", "Security", "lock|sentry|alarm|valet|guard"),
        new SignalCategory("motor", "signalDiff.cat.motor", "Motor", "motor|inverter|torque|rpm"),
        new SignalCategory("tire", "signalDiff.cat.tire", "Tire", "tpms|tire|pressure"),
        new SignalCategory("media", "signalDiff.cat.media", "Media", "media|audio|volume|playback"),
        new SignalCategory("safety", "signalDiff.cat.safety", "Safety", "airbag|seatbelt|fcw|aeb|safety"),
    };

    /// <summary>The category whose pattern matches <paramref name="signalName"/>, or null (first match wins).</summary>
    public static SignalCategory? Classify(string? signalName)
    {
        if (string.IsNullOrEmpty(signalName))
        {
            return null;
        }

        foreach (var category in All)
        {
            if (category.Matches(signalName))
            {
                return category;
            }
        }

        return null;
    }
}

/// <summary>The five datetime presets the compare bar offers — the native port of the web <c>DiffPresetId</c>.</summary>
public enum DiffPresetId
{
    /// <summary>Window A = one hour ago, Window B = now (web <c>now-vs-1h</c>).</summary>
    NowVs1h,

    /// <summary>Window A = one day ago, Window B = now (web <c>now-vs-1d</c>).</summary>
    NowVs1d,

    /// <summary>Window A = four hours ago, Window B = now (web <c>before-after-charge</c>).</summary>
    BeforeAfterCharge,

    /// <summary>Window A = 90 minutes ago, Window B = five minutes ago (web <c>last-drive</c>).</summary>
    LastDrive,

    /// <summary>Window A = one day ago, Window B = now (web <c>today-vs-yesterday</c>).</summary>
    TodayVsYesterday,
}

/// <summary>
/// One datetime preset — the native port of an entry in the web <c>DIFF_PRESETS</c> array. Carries the typed
/// <see cref="Id"/>, the stable <see cref="Wire"/> literal (web <c>id</c>), the i18n <see cref="LabelKey"/> +
/// English <see cref="DefaultLabel"/>, and the relative-time <see cref="Compute(DateTime)"/> the button applies.
/// Pure — the time maths is asserted against a fixed clock.
/// </summary>
/// <param name="Id">The typed preset id.</param>
/// <param name="Wire">The stable wire literal (web <c>id</c>), e.g. <c>now-vs-1h</c>.</param>
/// <param name="LabelKey">Web i18n key (web <c>labelKey</c>), e.g. <c>signalDiff.preset.nowVs1h</c>.</param>
/// <param name="DefaultLabel">English fallback label (web <c>defaultLabel</c>).</param>
public sealed record SignalDiffPreset(DiffPresetId Id, string Wire, string LabelKey, string DefaultLabel)
{
    /// <summary>Resolve the two window timestamps relative to <paramref name="now"/> (web <c>compute()</c>).</summary>
    public (DateTime AtA, DateTime AtB) Compute(DateTime now) => Id switch
    {
        DiffPresetId.NowVs1h => (now.AddHours(-1), now),
        DiffPresetId.NowVs1d => (now.AddDays(-1), now),
        DiffPresetId.BeforeAfterCharge => (now.AddHours(-4), now),
        DiffPresetId.LastDrive => (now.AddMinutes(-90), now.AddMinutes(-5)),
        DiffPresetId.TodayVsYesterday => (now.AddDays(-1), now),
        _ => (now, now),
    };
}

/// <summary>
/// The five datetime presets the compare bar offers, in display order — the native port of the web
/// <c>DIFF_PRESETS</c> export. Pure — unit-tested without a UI host.
/// </summary>
public static class SignalCompareControlsPresets
{
    /// <summary>The five presets in web declaration order.</summary>
    public static IReadOnlyList<SignalDiffPreset> All { get; } = new[]
    {
        new SignalDiffPreset(DiffPresetId.NowVs1h, "now-vs-1h", "signalDiff.preset.nowVs1h", "Now vs 1h ago"),
        new SignalDiffPreset(DiffPresetId.NowVs1d, "now-vs-1d", "signalDiff.preset.nowVs1d", "Now vs 1 day ago"),
        new SignalDiffPreset(DiffPresetId.BeforeAfterCharge, "before-after-charge", "signalDiff.preset.beforeAfterCharge", "Before vs after last charge"),
        new SignalDiffPreset(DiffPresetId.LastDrive, "last-drive", "signalDiff.preset.lastDrive", "Last drive start vs end"),
        new SignalDiffPreset(DiffPresetId.TodayVsYesterday, "today-vs-yesterday", "signalDiff.preset.todayVsYesterday", "Today vs yesterday (same time)"),
    };

    /// <summary>The preset with the given id (always present for a valid enum value).</summary>
    public static SignalDiffPreset Get(DiffPresetId id) => All.First(p => p.Id == id);
}

/// <summary>
/// Pure ports of the web datetime helpers <c>toLocalDatetimeInput</c> and <c>isoOrEmpty</c> plus the inverse
/// parse the native pickers need. Kept WinUI-free so the round-trip is asserted headlessly.
/// </summary>
public static class SignalCompareControlsTime
{
    private const string LocalFormat = "yyyy-MM-ddTHH:mm";

    /// <summary>
    /// Format a local datetime as the <c>datetime-local</c> input value <c>yyyy-MM-ddTHH:mm</c> (web
    /// <c>toLocalDatetimeInput</c>). Seconds and zone are intentionally dropped to match the HTML control.
    /// </summary>
    public static string ToLocalDatetimeInput(DateTime local) =>
        local.ToString(LocalFormat, CultureInfo.InvariantCulture);

    /// <summary>
    /// Convert a <c>datetime-local</c> value to a UTC ISO-8601 instant, or the empty string for blank/invalid
    /// input (web <c>isoOrEmpty</c>: <c>new Date(local).toISOString()</c>). The value is read as local time and
    /// normalized to UTC with millisecond precision and a trailing <c>Z</c>, matching JavaScript's
    /// <c>toISOString()</c>.
    /// </summary>
    public static string IsoOrEmpty(string? localValue)
    {
        if (string.IsNullOrEmpty(localValue))
        {
            return string.Empty;
        }

        if (!DateTime.TryParse(
                localValue,
                CultureInfo.InvariantCulture,
                DateTimeStyles.AssumeLocal,
                out var parsed))
        {
            return string.Empty;
        }

        return parsed.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss.fffZ", CultureInfo.InvariantCulture);
    }

    /// <summary>Parse a <c>datetime-local</c> value back to a local <see cref="DateTime"/> (false when blank/invalid).</summary>
    public static bool TryParseLocalInput(string? localValue, out DateTime value)
    {
        value = default;
        return !string.IsNullOrEmpty(localValue)
            && DateTime.TryParse(localValue, CultureInfo.InvariantCulture, DateTimeStyles.None, out value);
    }
}

/// <summary>
/// The render-time data model the controlled compare bar binds to — the native analogue of the web
/// <c>SignalCompareControlsProps</c> data fields (<c>atA</c>, <c>atB</c>, <c>search</c>, <c>category</c>). The
/// component is purely controlled: it holds the current selection and raises change events the host applies; it
/// performs no fetching and mutates nothing itself. The web <c>topSlot</c> / <c>className</c> are view concerns
/// (a slot element and styling), not data, so they live on the WinUI view rather than this model. Pure data — the
/// projection is unit-tested without a UI host.
/// </summary>
/// <param name="AtA">The Window A <c>datetime-local</c> value (web <c>atA</c>).</param>
/// <param name="AtB">The Window B <c>datetime-local</c> value (web <c>atB</c>).</param>
/// <param name="Search">The signal-name filter text (web <c>search</c>).</param>
/// <param name="Category">The active category id, or null for none (web <c>category</c>).</param>
public sealed record SignalCompareControlsModel(
    string AtA,
    string AtB,
    string Search,
    string? Category)
{
    /// <summary>The initial model: both windows and the filter empty, no category selected.</summary>
    public static SignalCompareControlsModel Empty { get; } =
        new(string.Empty, string.Empty, string.Empty, null);
}

/// <summary>A render-ready preset button (web <c>DIFF_PRESETS.map</c> entry): typed id, wire literal, localized label.</summary>
public sealed record PresetButton(DiffPresetId Id, string Wire, string Label);

/// <summary>A render-ready category chip (web <c>CATEGORY_PREFIXES.map</c> entry): id, localized label, active flag.</summary>
public sealed record CategoryChip(string Id, string Label, bool Active);

/// <summary>
/// The fully projected, render-ready view of the compare bar for one input model — the native analogue of what
/// the web <c>SignalCompareControls</c> renders. Holds the two window field labels (+ their help body / aria
/// copy), the presets-row label and the five preset buttons, the signal-name filter hint, the eight category chips
/// (each carrying its <see cref="CategoryChip.Active"/> flag), the clear-affordance visibility + label, the
/// echoed selection (so the view sets the pickers / filter), and the surface automation name. Pure data so every
/// branch is asserted headlessly.
/// </summary>
public sealed record SignalCompareControlsDisplay(
    string WindowALabel,
    string WindowBLabel,
    string SnapshotHelp,
    string SnapshotHelpAria,
    string DiffHelp,
    string DiffHelpAria,
    string PresetsLabel,
    IReadOnlyList<PresetButton> Presets,
    string FilterHint,
    IReadOnlyList<CategoryChip> Categories,
    bool ShowClear,
    string ClearLabel,
    string AtA,
    string AtB,
    string Search,
    string AutomationName);

/// <summary>
/// Pure projection from a <see cref="SignalCompareControlsModel"/> to its <see cref="SignalCompareControlsDisplay"/>
/// — the native port of <c>web/src/features/telemetry/components/SignalCompareControls.tsx</c>. It resolves the
/// two window labels and their help copy, builds the five preset buttons and the eight category chips (marking the
/// one whose id equals the model's <c>category</c> as active, exactly as the web compares
/// <c>category === c.id</c>), decides whether the "Clear" affordance shows (the web renders it only when a
/// category is selected), echoes the current selection, and resolves every string through the i18n facade. No
/// WinUI types — unit-tested without a UI host.
/// </summary>
public static class SignalCompareControlsProjection
{
    /// <summary>i18n key prefix every web key is namespaced under in the resource catalog.</summary>
    public const string KeyPrefix = "translation.";

    private const string WindowAKey = "translation.signalDiff.windowA";
    private const string WindowBKey = "translation.signalDiff.windowB";
    private const string SnapshotHelpKey = "translation.help.signal.snapshot";
    private const string SnapshotHelpAriaKey = "translation.help.signal.snapshot.aria";
    private const string DiffHelpKey = "translation.help.signal.diff";
    private const string DiffHelpAriaKey = "translation.help.signal.diff.aria";
    private const string PresetsLabelKey = "translation.signalDiff.presetsLabel";
    private const string FilterHintKey = "translation.signalDiff.filterPlaceholder"; // parity:allow web i18n key signalDiff.filterPlaceholder
    private const string ClearKey = "translation.signalDiff.clearCategory";

    private const string WindowAFallback = "Window A";
    private const string WindowBFallback = "Window B";
    private const string SnapshotHelpFallback =
        "A snapshot is a point-in-time view of every signal value at a single timestamp. " +
        "Falls back to signal_log within the last 30 days when the live layer doesn\u2019t have it.";
    private const string SnapshotHelpAriaFallback = "More info about signal snapshots";
    private const string DiffHelpFallback =
        "Server-side comparison between two snapshots. Unchanged signals are omitted from the result to reduce noise.";
    private const string DiffHelpAriaFallback = "More info about signal diffs";
    private const string PresetsLabelFallback = "Quick presets:";
    private const string FilterHintFallback = "Filter signals\u2026";
    private const string ClearFallback = "Clear";

    /// <summary>Project <paramref name="model"/> into a render-ready display using the i18n facade.</summary>
    /// <param name="model">The render-time data model (the web data props).</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    public static SignalCompareControlsDisplay Project(SignalCompareControlsModel model, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        string windowA = localizer.GetString(WindowAKey, WindowAFallback);
        string windowB = localizer.GetString(WindowBKey, WindowBFallback);
        string presetsLabel = localizer.GetString(PresetsLabelKey, PresetsLabelFallback);
        string filterHint = localizer.GetString(FilterHintKey, FilterHintFallback);
        string clearLabel = localizer.GetString(ClearKey, ClearFallback);

        var presets = BuildPresets(localizer);
        var categories = BuildCategories(model.Category, localizer);
        bool showClear = !string.IsNullOrEmpty(model.Category);

        return new SignalCompareControlsDisplay(
            WindowALabel: windowA,
            WindowBLabel: windowB,
            SnapshotHelp: localizer.GetString(SnapshotHelpKey, SnapshotHelpFallback),
            SnapshotHelpAria: localizer.GetString(SnapshotHelpAriaKey, SnapshotHelpAriaFallback),
            DiffHelp: localizer.GetString(DiffHelpKey, DiffHelpFallback),
            DiffHelpAria: localizer.GetString(DiffHelpAriaKey, DiffHelpAriaFallback),
            PresetsLabel: presetsLabel,
            Presets: presets,
            FilterHint: filterHint,
            Categories: categories,
            ShowClear: showClear,
            ClearLabel: clearLabel,
            AtA: model.AtA,
            AtB: model.AtB,
            Search: model.Search,
            AutomationName: BuildAutomationName(windowA, windowB, presetsLabel, categories));
    }

    /// <summary>The category id toggled when chip <paramref name="chipId"/> is tapped (web <c>category === id ? null : id</c>).</summary>
    public static string? ToggleCategory(string? current, string chipId)
    {
        ArgumentException.ThrowIfNullOrEmpty(chipId);
        return string.Equals(current, chipId, StringComparison.Ordinal) ? null : chipId;
    }

    // Web parity: DIFF_PRESETS.map(p => ({ id: p.id, label: t(p.labelKey, p.defaultLabel) })).
    private static List<PresetButton> BuildPresets(ILocalizer localizer)
    {
        var buttons = new List<PresetButton>(SignalCompareControlsPresets.All.Count);
        foreach (var preset in SignalCompareControlsPresets.All)
        {
            buttons.Add(new PresetButton(
                preset.Id,
                preset.Wire,
                localizer.GetString(KeyPrefix + preset.LabelKey, preset.DefaultLabel)));
        }

        return buttons;
    }

    // Web parity: CATEGORY_PREFIXES.map(c => ({ id: c.id, label: t(c.labelKey, c.defaultLabel),
    // active: category === c.id })).
    private static List<CategoryChip> BuildCategories(string? active, ILocalizer localizer)
    {
        var chips = new List<CategoryChip>(SignalCompareControlsCategories.All.Count);
        foreach (var category in SignalCompareControlsCategories.All)
        {
            chips.Add(new CategoryChip(
                category.Id,
                localizer.GetString(KeyPrefix + category.LabelKey, category.DefaultLabel),
                string.Equals(active, category.Id, StringComparison.Ordinal)));
        }

        return chips;
    }

    private static string BuildAutomationName(
        string windowA,
        string windowB,
        string presetsLabel,
        IReadOnlyList<CategoryChip> categories)
    {
        var active = categories.FirstOrDefault(c => c.Active);
        string baseName = string.Create(
            CultureInfo.CurrentCulture,
            $"{windowA}. {windowB}. {presetsLabel}");

        return active is null
            ? baseName
            : string.Create(CultureInfo.CurrentCulture, $"{baseName} {active.Label}");
    }
}

/// <summary>
/// PII-safe diagnostics for the <c>SignalCompareControls</c> surface (P1/S11 diagnostics contract). Records only
/// the operational <c>view.opened</c> event with the surface slug — never the selected timestamps, filter text or
/// category — so a diagnostics line can never leak what an operator compared. Thread-safe.
/// </summary>
public sealed class SignalCompareControlsDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">An optional operational-only line sink (no user data is ever passed).</param>
    public SignalCompareControlsDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=SignalCompareControls</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={SignalCompareControlsRegistration.Slug}");
    }
}

/// <summary>
/// Canonical metadata for the <c>SignalCompareControls</c> feature surface — the native mirror of the web
/// component at <c>web/src/features/telemetry/components/SignalCompareControls.tsx</c>.
/// </summary>
public static class SignalCompareControlsRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event.</summary>
    public const string Slug = "SignalCompareControls";
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/SignalCompareControls/SignalCompareControlsModel.kt
// Pure, framework-free model + projection for the SignalCompareControls feature view — the native analogue of
// everything the web component derives before returning JSX
// (web/src/features/telemetry/components/SignalCompareControls.tsx). No Compose, no Android, no HTTP: every
// declaration here is unit-tested off-device in the :android:testReleaseUnitTest gate, keeping the composable a
// thin render layer.
//
// SignalCompareControls is, in the web source's own words, "Pure controls (no data fetching, no diff table)":
// it receives the two compare windows (`atA` / `atB` as `datetime-local` strings), the signal `search`, and the
// active `category`, plus their change callbacks, and is mounted by both SignalDiffPage and SignalsWorkspacePage.
// The one data source the web binds is `useTranslation`. So, exactly as the sibling WeekSelector / AddWidgetButton
// presentational ports document, the loading / empty / error / stale / offline lifecycle lives on the OWNING
// page, not here; the conditional branches the web source itself defines — the optional top slot and the
// category-clear affordance shown only while a category is active — are the complete branch set this surface
// renders, alongside the always-visible windows, presets, filter and chips.
//
// This file owns the pure parts the web component exports so the pages can also drive their server-side filter
// strings: the eight category prefixes with their verbatim regex matchers (web `CATEGORY_PREFIXES`), the five
// quick presets with their relative-window math (web `DIFF_PRESETS`), the `datetime-local` <-> ISO helpers (web
// `toLocalDatetimeInput` / `isoOrEmpty`), the category toggle decision (web `category === c.id ? null : c.id`),
// the `t(key, default)` resolver for the two aria strings the i18n catalog does not define, and the PII-safe
// diagnostics contract (P1/S11).
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/SignalCompareControls — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.signalcomparecontrols

import io.teslasync.shared.core.diagnostics.Logger
import java.time.Duration
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * The PII-safe diagnostics contract this surface emits (P1/S11). Carries only the surface [SLUG] — never a
 * window timestamp, signal name or category, any of which could narrow which vehicle/window an operator was
 * inspecting — so a diagnostics line can never leak the comparison an operator set up.
 */
object SignalCompareControlsDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = "SignalCompareControls"

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}

/**
 * One signal-name category prefix — the native mirror of a single web `CATEGORY_PREFIXES` entry. [id] is the
 * verbatim web id (the controlled-selection token the pages persist), and [matches] reproduces the web entry's
 * case-insensitive `matches(name)` regex so a host page can fold the live signal list into the same buckets the
 * chips toggle. Declaration order IS the web array order, so [ALL] renders the same chip sequence.
 */
enum class DiffCategory(
    val id: String,
    private val pattern: Regex,
) {
    Battery("battery", Regex("battery|charge|soc|range|kwh", RegexOption.IGNORE_CASE)),
    Drive("drive", Regex("speed|odometer|gear|drive|brake|throttle|steering", RegexOption.IGNORE_CASE)),
    Climate("climate", Regex("climate|hvac|cabin|seat|temp", RegexOption.IGNORE_CASE)),
    Security("security", Regex("lock|sentry|alarm|valet|guard", RegexOption.IGNORE_CASE)),
    Motor("motor", Regex("motor|inverter|torque|rpm", RegexOption.IGNORE_CASE)),
    Tire("tire", Regex("tpms|tire|pressure", RegexOption.IGNORE_CASE)),
    Media("media", Regex("media|audio|volume|playback", RegexOption.IGNORE_CASE)),
    Safety("safety", Regex("airbag|seatbelt|fcw|aeb|safety", RegexOption.IGNORE_CASE)),
    ;

    /** Whether [signalName] belongs to this category — web `matches: (n) => /.../i.test(n)`. */
    fun matches(signalName: String): Boolean = pattern.containsMatchIn(signalName)

    companion object {
        /** The chip order — the native mirror of the exported web `CATEGORY_PREFIXES` array. */
        val ALL: List<DiffCategory> = entries.toList()

        /** The category whose [id] equals the argument, or `null` (e.g. for the cleared selection). */
        fun fromId(id: String?): DiffCategory? = entries.firstOrNull { it.id == id }
    }
}

/**
 * A render-ready compare window pair — the two `datetime-local` strings a preset resolves to, matching the web
 * `applyPreset` which feeds `onChangeA(toLocalDatetimeInput(a))` / `onChangeB(toLocalDatetimeInput(b))`. Pure
 * data so the preset math is unit-tested without a UI host.
 */
data class DiffWindow(
    val atA: String,
    val atB: String,
)

/**
 * One quick datetime preset — the native mirror of a single web `DIFF_PRESETS` entry. [id] is the verbatim web
 * id; the offsets are how far before "now" each window sits, reproducing each web `compute()` exactly (Window B
 * is "now" for every preset except Last drive, which ends five minutes back). Declaration order IS the web array
 * order, so [ALL] renders the same button sequence.
 */
enum class DiffPreset(
    val id: String,
    private val offsetA: Duration,
    private val offsetB: Duration,
) {
    NowVs1h("now-vs-1h", Duration.ofHours(1), Duration.ZERO),
    NowVs1d("now-vs-1d", Duration.ofDays(1), Duration.ZERO),
    BeforeAfterCharge("before-after-charge", Duration.ofHours(4), Duration.ZERO),
    LastDrive("last-drive", Duration.ofMinutes(90), Duration.ofMinutes(5)),
    TodayVsYesterday("today-vs-yesterday", Duration.ofDays(1), Duration.ZERO),
    ;

    /**
     * Resolves this preset's window pair relative to [nowMillis], formatting each end in [zone] — web
     * `compute()` followed by `toLocalDatetimeInput`. [nowMillis] / [zone] are injected (never read from a global
     * clock) so the math is deterministic under test.
     */
    fun compute(
        nowMillis: Long,
        zone: ZoneId,
    ): DiffWindow {
        val now = Instant.ofEpochMilli(nowMillis)
        return DiffWindow(
            atA = SignalCompareTime.toLocalDatetimeInput(now.minus(offsetA).toEpochMilli(), zone),
            atB = SignalCompareTime.toLocalDatetimeInput(now.minus(offsetB).toEpochMilli(), zone),
        )
    }

    companion object {
        /** The button order — the native mirror of the exported web `DIFF_PRESETS` array. */
        val ALL: List<DiffPreset> = entries.toList()

        /** The preset whose [id] equals the argument, or `null` for an unrecognised token. */
        fun fromId(id: String): DiffPreset? = entries.firstOrNull { it.id == id }
    }
}

/**
 * The `datetime-local` <-> ISO helpers — the native mirror of the web `toLocalDatetimeInput` / `isoOrEmpty`
 * free functions. The window strings use the exact HTML `datetime-local` shape (`yyyy-MM-ddTHH:mm`) so the
 * controlled-prop contract matches the web component one-for-one and a host can drive the same server filter.
 */
object SignalCompareTime {
    private const val LOCAL_PATTERN = "yyyy-MM-dd'T'HH:mm"
    private const val DISPLAY_PATTERN = "yyyy-MM-dd HH:mm"
    private val LOCAL_FORMAT: DateTimeFormatter = DateTimeFormatter.ofPattern(LOCAL_PATTERN, Locale.ROOT)
    private val DISPLAY_FORMAT: DateTimeFormatter = DateTimeFormatter.ofPattern(DISPLAY_PATTERN, Locale.ROOT)

    /**
     * Formats [epochMillis] (an absolute instant) as a `datetime-local` string in [zone] — web
     * `toLocalDatetimeInput`.
     */
    fun toLocalDatetimeInput(
        epochMillis: Long,
        zone: ZoneId,
    ): String =
        Instant
            .ofEpochMilli(epochMillis)
            .atZone(zone)
            .toLocalDateTime()
            .format(LOCAL_FORMAT)

    /** Formats an already-local [value] as a `datetime-local` string — the picker-confirm path. */
    fun toLocalDatetimeInput(value: LocalDateTime): String = value.format(LOCAL_FORMAT)

    /**
     * Parses a `datetime-local` string back to a [LocalDateTime], or `null` when blank or malformed — used to
     * seed the date/time pickers from the current window value. Mirrors the web component reading `value={atA}`.
     */
    fun parseLocalDatetime(localValue: String): LocalDateTime? {
        if (localValue.isBlank()) return null
        return runCatching { LocalDateTime.parse(localValue, LOCAL_FORMAT) }.getOrNull()
    }

    /**
     * Converts a `datetime-local` [localValue] (interpreted in [zone]) into a UTC ISO-8601 instant string, or
     * `""` when blank/invalid — the native mirror of web `isoOrEmpty` (`new Date(local).toISOString()`), the
     * exact server-filter string the owning pages send.
     */
    fun isoOrEmpty(
        localValue: String,
        zone: ZoneId,
    ): String {
        val parsed = parseLocalDatetime(localValue) ?: return ""
        return parsed.atZone(zone).toInstant().toString()
    }

    /**
     * The human-friendly field text for a window [localValue]: the parsed datetime rendered as
     * `yyyy-MM-dd HH:mm`, or [emptyLabel] when the window is unset — so the tap-to-pick field is never an empty
     * box (the web shows the browser's native `datetime-local` glyph for the same empty state).
     */
    fun displayLabel(
        localValue: String,
        emptyLabel: String,
    ): String = parseLocalDatetime(localValue)?.format(DISPLAY_FORMAT) ?: emptyLabel
}

/**
 * The category-chip toggle decision — the native mirror of web `onCategoryChange(category === c.id ? null : c.id)`:
 * tapping the active chip clears the selection, tapping any other chip selects it. Pure so it is unit-tested
 * without a UI host.
 */
fun toggleCategory(
    current: String?,
    clicked: String,
): String? = if (current == clicked) null else clicked

/** Resource name (absent ⇒ [SignalCompareDefaults.SNAPSHOT_ARIA]) for the Window A help trigger's accessible name. */
const val KEY_SNAPSHOT_ARIA: String = "translation_help_signal_snapshot_aria"

/** Resource name (absent ⇒ [SignalCompareDefaults.DIFF_ARIA]) for the Window B help trigger's accessible name. */
const val KEY_DIFF_ARIA: String = "translation_help_signal_diff_aria"

/** Resource name (absent ⇒ [SignalCompareDefaults.PICK_WINDOW]) for the empty-window tap-to-pick label. */
const val KEY_PICK_WINDOW: String = "translation_signalDiff_pickWindow"

/**
 * Native fallback microcopy reproducing i18next's "return the default when the key is absent" behaviour for the
 * three strings the web resolves via `t(key, default)` whose keys the generated catalog (P1/S10) does not define:
 * the two help-trigger aria names (web `ariaLabel={t('help.signal.*.aria', { defaultValue: '…' })}`) and the
 * empty-window tap-to-pick label (the web's native `datetime-local` input shows the browser's own empty glyph;
 * a tap-to-pick field needs an explicit, localized label so the control is never an empty box).
 */
object SignalCompareDefaults {
    /** web `help.signal.snapshot.aria` default. */
    const val SNAPSHOT_ARIA: String = "More info about signal snapshots"

    /** web `help.signal.diff.aria` default. */
    const val DIFF_ARIA: String = "More info about signal diffs"

    /** Empty-window tap-to-pick label. */
    const val PICK_WINDOW: String = "Select date & time"
}

/**
 * Reproduces i18next's `t(key, default)` against the native i18n facade: returns [lookup] of [resourceName] when
 * it resolves to a non-blank string, otherwise [fallback]. [lookup] is a thin seam over the Android string
 * catalog in production (an optional by-name read) and a map in tests, keeping the resolve-or-fallback decision
 * pure and unit-tested.
 */
fun resolveOptional(
    lookup: (String) -> String?,
    resourceName: String,
    fallback: String,
): String = lookup(resourceName)?.takeIf { it.isNotBlank() } ?: fallback
